In [1]:
import pandas as pd
import numpy as np
import statsmodels.api as sm
from statsmodels.api import add_constant
from statsmodels.sandbox.regression.predstd import wls_prediction_std

In [2]:
df = pd.read_excel("./data-table-B3.XLS")

WARNING *** file size (10909) not 512 + multiple of sector size (512)
WARNING *** OLE2 inconsistency: SSCS size is 0 but SSAT size is non-zero


In [3]:
df = df.sample(n=22)

In [4]:
df.head()

y     x1   x2     x3   x4    x5  x6  x7     x8    x9   x10  x11
11  36.50   85.3   80   83.0  8.5  3.89   2   4  160.6  62.2  2009    0
7   21.47  262.0  110  200.0  8.5  2.56   2   3  179.3  65.4  3180    1
20  23.54  231.0  110  175.0  8.0  2.56   2   3  179.3  65.4  3050    1
31  16.50  360.0  165  255.0  8.5  2.73   4   3  185.2  69.0  3660    1
23  31.90   96.9   75   83.0  9.0  4.30   2   5  165.2  61.8  2275    0

A. Fit a multiple linear regression model relatmg gasoline mileage y (miles per
gallon) to engine displacement x1 and the number of carburetor barrels x6 .

In [55]:
X = df[['x1', 'x6']]
y = df['y']

reg = sm.OLS(y, X)
fit = reg.fit()
print(fit.params)

x1   -0.003435
x6    6.575100
dtype: float64


b. Construct the analysis - of - variance table and test for signifi cance of regression.

In [56]:
fit.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                      y   R-squared (uncentered):                   0.689
Model:                            OLS   Adj. R-squared (uncentered):              0.658
Method:                 Least Squares   F-statistic:                              22.18
Date:                Wed, 16 Sep 2020   Prob (F-statistic):                    8.39e-06
Time:                        14:46:14   Log-Likelihood:                         -85.662
No. Observations:                  22   AIC:                                      175.3
Df Residuals:                      20   BIC:                                      177.5
Df Model:                           2                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
x1            -0.0034      0.029     -0.119      0.907      -0.064       0.057
x6             6.5751      3.142      2.092      0.049       0.020      13.130
==============================================================================
Omnibus:                        6.119   Durbin-Watson:                   2.357
Prob(Omnibus):                  0.047   Jarque-Bera (JB):                1.731
Skew:                          -0.042   Prob(JB):                        0.421
Kurtosis:                       1.628   Cond. No.                         366.
==============================================================================

Notes:
[1] R² is computed without centering (uncentered) since the model does not contain a constant.
[2] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

 c. Calculate R2 and RAdj 2 for this model. Compare this to the R2 and the RAdj 2 for the simple linear regression model relating mileage to engine displacement in Problem 2.4.

<font color="blue">Indicated above</font>

R-sq = 0.689 <br>
Adj Rsq = 0.658

<font color="blue">Problem 2.4 below</font>
The R-squared and adj R-sq had increased with the additional regressors.

In [57]:
df24 = df.dropna()

In [59]:
# 2.4 uses all regressors
X24 = df24.loc[:, df24.columns !='y']
y24 = df24['y']

reg24 = sm.OLS(y24, X24)
fit24 = reg24.fit()
print(fit24.summary())

                                 OLS Regression Results                                
Dep. Variable:                      y   R-squared (uncentered):                   0.989
Model:                            OLS   Adj. R-squared (uncentered):              0.977
Method:                 Least Squares   F-statistic:                              83.52
Date:                Wed, 16 Sep 2020   Prob (F-statistic):                    2.51e-08
Time:                        14:46:45   Log-Likelihood:                         -46.001
No. Observations:                  21   AIC:                                      114.0
Df Residuals:                      10   BIC:                                      125.5
Df Model:                          11                                                  
Covariance Type:            nonrobust                                                  
                 coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------

 d. Find a 95% CI for β1 .

In [60]:
[-0.064, 0.057]

# fit.conf_int(alpha=0.05, cols=(1)

[-0.064, 0.057]

 e. Compute the t statistics for testing H0 : β1 = 0 and H0 : β6 = 0. What conclusions can you draw?

In [61]:
hypotheses = 'x1 = 0'
fit.t_test(hypotheses)

<class 'statsmodels.stats.contrast.ContrastResults'>
                             Test for Constraints                             
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
c0            -0.0034      0.029     -0.119      0.907      -0.064       0.057

In [62]:
hypotheses6 = 'x6 = 0'
fit.t_test(hypotheses6)

<class 'statsmodels.stats.contrast.ContrastResults'>
                             Test for Constraints                             
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
c0             6.5751      3.142      2.092      0.049       0.020      13.130

Reject null for x6

 f. Find a 95% CI on the mean gasoline mileage when x1 = 275 in. 3 and x6 = 2 barrels. 

In [65]:
c = [275, 2]
result = fit.get_prediction(c)
result.conf_int(alpha=0.05)

array([[ 6.56552345, 17.8457149 ]])

g. Find a 95% prediction interval for a new observation on gasoline mileage when x1 = 257 in. 3 and x6 = 2 barrels.

In [66]:
sdev1, lower_pred1, upper_pred1 = wls_prediction_std(fit, c, alpha=0.95)
[lower_pred1, upper_pred1]

[array([11.39608493]), array([13.01515342])]